client/src/includes/stimulus.ts

```diff
@@ -0,0 +1,201 @@
+ * })
+ *
+ */
+export const createController = (
+  controllerDefinition: ControllerObjectDefinition = {},
+): typeof Controller => {
+  class NewController<X extends Element> extends Controller<X> {}
+
+  // set up static values
+  Object.entries(controllerDefinition.STATIC || {}).forEach(([key, value]) => {
+    NewController[key] = value;
+  });
+
+  // set up class methods
+  Object.entries(controllerDefinition)
+    .filter(([key]) => !['STATIC'].includes(key))
+    .map(([key]) => ({
+      key,
+      property:
+        Object.getOwnPropertyDescriptor(controllerDefinition, key) || noop,
+    }))
+    .forEach(({ key, property }) => {
+      Object.defineProperty(NewController.prototype, key, property);
+    });
+
+  return NewController;
+};
+
+/**
+ * Dispatches events and adds event listeners for documented API surface without relying
+ * on exposing the application instance or adding more globals.
+ *
+ * @example - within Wagtail for custom stimulus controllers
+ * ```python
+ * @hooks.register('insert_global_admin_js')
+ * def global_admin_js():
+ *     return mark_safe(
+ *         """
+ * <script type="module">
+ * import { Controller } from "https://unpkg.com/@hotwired/stimulus/dist/stimulus.js";
+ *
+ * class HelloController extends Controller {
+ *   static targets = ["name"];
+ *
+ *   connect() {
+ *     console.log("connected");
+ *   }
+ * }
+ *
+ * document.addEventListener(
+ *   "wagtail:stimulus-ready",
+ *   ({ detail: { register } }) => {
+ *     register("hello", HelloController);
+ *   },
+ *   { once: true }
+ * );
+ * </script>
+ *         """
+ *     )
+ * ```
+ */
+const setupApplication = (application: Application) => {
+  /**
+   * Allows for safe controller registration with logging, with more explicit object
+   * param for controller & identifier so that additional arguments can be added later
+   * if needed.
+   */
+  const register = (
+    identifier?: string,
+    controller?: ControllerConstructor,
+  ) => {
+    if (!controller || !identifier) {
+      application.logDebugActivity(
+        'registration failed',
+        controller ? 'no identifier' : 'no controller',
+        { controller, identifier },
+      );
+      return;
+    }
+    application.load({ identifier, controllerConstructor: controller });
+    application.logDebugActivity('registered', identifier);
+  };
+
+  /**
+   * Dispatch event for stimulus being initialised.
+   * Most code will not be able to listen to this event,
+   * setting up for any future fallbacks/provision of application or
+   * cancelling/modifying core controllers.
+   */
+  document.dispatchEvent(new CustomEvent('wagtail:stimulus-init'));
+
+  /**
+   * Add event listener to allow any custom code to trigger debug mode easily.
+   */
+  window.addEventListener('wagtail:stimulus-enable-debug', () => {
+    // eslint-disable-next-line no-param-reassign
+    application.debug = true;
+    application.logDebugActivity('debug enabled', 'logDebugActivity');
+  });
```

Comment:
This strikes me as a good example of where exposing the application on the window object would be much simpler. "`Stimulus.debug = true` in the console" is what the Stimulus docs explicitly recommend.

Reply:
OK - on the `window.Stimulus` discussion - I think I need to heed the advice here and give this a go. I just want to double check - we are ok with adding yet another global in our quest to reduce globals across Wagtail. This new global being the exception :) 

If we are, can we keep this focused and attach the other 'things' we need to the instance (or as a static method).

We would provide the application instance as `window.Stimulus` and also provide the following;

* `window.Stimulus.createController` - the helper function that creates a controller from a plain object for those that do not want to think about classes or build tooling.
* `window.Stimulus.Controller` - this would be made available so that those who want to create a class can easily extend this with their own class (note: this will break for any Node build system based code that is NOT using ES6 or above as their compile target - we can document this of course).

If we want 'control' over behaviour for things like `register` we can override that class method and inject custom behaviour easily in the future.

We lose the ability to be strict about what 'comes in' to our Stimulus instance (there are always way that devs can do this but its harder if its not at the global). They will be able to `stop` the application instance which would break all widgets globally, unload any controllers, use the `getControllerForElementAndIdentifier` to directly access controller instances more easily (this seems like it's ok but can lead to more code being tied to the controller name).

These may be low risk (if developers break their own code, that is up to them), but it is a risk nonetheless.

If projects want to use their own isolated Stimulus instance, they of course will need to use a different global name.

It will be harder to change/rename this global once it's there.

An alternative approach is to expose a faux Stimulus instance which only exposes the things we are ready to support as a formal API (even if undocumented).

Reply:
This will be updated soon so I will resolve this issue.

API will be via the `constructor` so;

`window.Stimulus.constructor.Controller`  (base class to extend)
`window.Stimulus.constructor.createController` (from object)

so full usage would be something like

```js
// create a controller from object and register it
(() => {
  const WordCountController = window.Stimulus.constructor.createController({
    STATIC: {
      values: { max: { default: 10, type: Number } },
    },
    connect() {
      console.log(`MaxValue: ${this.maxValue} - connected to:`, this.element);
    },
  });
  window.Stimulus.register('example-a', WordCountController);
})();
```

OR

```js
// create a controller class and register it (will not work if using Node tooling that compiles down to ES5)
(() => {
  class WordCountController extends window.stimulus.constructor.Controller {
    static values = { max: { default: 10, type: Number } };
    connect() {
      console.log(`MaxValue: ${this.maxValue} - connected to:`, this.element);
    }
  }
  window.Stimulus.register('example-a', WordCountController);
})();
```